plugin/src/main/java/io/jenkins/plugins/casc/SecretSourceResolver.java

```diff
@@ -0,0 +1,48 @@
+package io.jenkins.plugins.casc;
+
+import static io.vavr.API.unchecked;
+
+import org.bigtesting.interpolatd.Interpolator;
+
+import java.util.Optional;
+import java.util.stream.Stream;
+
+public class SecretSourceResolver {
+    private static final String enclosedBy = "${";
+    private static final String enclosedIn = "}";
+    private static final String escapedWith = "^";
+    private static final String defaultDelimiter = ":-";
+
+    public static String resolve(ConfigurationContext context, String toInterpolate) {
+        return interpolator(context).interpolate(toInterpolate, "");
+    }
+
+    private static Interpolator<String> interpolator(ConfigurationContext context) {
+        Interpolator<String> interpolator = new Interpolator<>();
+        interpolator.when().enclosedBy(enclosedBy).and(enclosedIn).handleWith((captured, argument) -> handle(context, captured));
+        interpolator.escapeWith(escapedWith);
+        return interpolator;
+    }
+
+    private static String handle(ConfigurationContext context, String captured) {
+        return reveal(context, captured)
+                .map(Optional::of)
+                .orElse(defaultValue(captured))
+                .orElse("");
```

Comment:
Doesn’t it currently leave the value alone if it can’t find a replacement I.e ‘${MY_SECRET}’ will be the final value not empty string?

No strong feeling either way on the correctness though

Reply:
My original implementation did leave the value alone when it couldn't find a replacement (`${FOO}` would yield `FOO`), but `ArtifactManagerS3Test` failed. After doing some debugging, I saw that the previous code would return empty string when that's the case, so I made it align with the old implementation.